### PR TITLE
Cloud API: Add support for Google Cloud Platform as a target

### PR DESCRIPTION
**Pull request by thozza**
- Add internal Google Cloud Platform API supporting uploading images to GCP and sharing them with specified accounts.
- Add CLI tool `osbuild-upload-gcp` using the internal GCP API.
- Extend Cloud API to support GCP as an upload target.
- Extend worker job implementation to be able to handle GCP as an upload target.
- Rework `api.sh` to allow integration testing for multiple cloud providers.
- Extend `api.sh` to also test GCP.

This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
  - Integration test `api.sh` has been extended to test also GCP integration.
- [x] adequate documentation informing people about the change
  - Updated `test/README.md` to note integration testing of Cloud API
  - Added NEWS entry for the new functionality.

### CODE REVIEW

**Comment by thozza** [comment]
I still need to verify that the changes don't affect `cockpit-composer`, but feel free to review in the meantime. Thanks!

File: internal/cloudapi/openapi.yml
Comment (teg) [inline internal/cloudapi/openapi.yml:268]: Any particular reason why/why not?

File: internal/cloudapi/server.go
Comment (teg) [inline internal/cloudapi/server.go:238]: I would say: either let the caller decide the name, and require them to choose a unique one, or don't and use the random string as below.

File: tools/provision.sh
Comment (teg) [inline tools/provision.sh:18]: Maybe we should have a separate `worker` subdirectory which is copied to `/etc/osbuild-worker`?

File: test/cases/api.sh
Comment (teg) [inline test/cases/api.sh:414]: This should probably also instantiate / boot the image like for AWS, right? For a follow-up?

File: internal/cloudapi/server.go
Comment (teg) [inline internal/cloudapi/server.go:262]: @Gundersanne this won't be a problem for image-builder, I assume? This value is currently ignored?

File: test/README.md
Comment (teg) [inline test/README.md:220]: Love this documentation! Another way to test this would be to use `docker-compose up --build` (introduced by @achilleas-k) rather than building/installing on your local machine. To make that really smooth, we may want to introduce a third container in which to actually run the test scripts, but not sure if that's necessary.
No need to change this or look into the docker stuff now though, just wanted to mention it in case you didn't know about it.

**Review by teg — APPROVED** [review]
This is really great! Thanks so much @thozza!

**Comment by teg** [comment]
Seems we need some creds in Jenkins though. Do we have some we can use, or are we blocked on this?

File: internal/cloudapi/server.go
Comment (Gundersanne) [reply internal/cloudapi/server.go:262]: This is fine yea, it's currently ignored as you say.

File: internal/cloudapi/openapi.yml
Comment (thozza) [reply internal/cloudapi/openapi.yml:268]: Honestly, this is one thing that I wanted to discuss as part of the PR. The "os type" is used when importing an image into Compute Node from previously uploaded image from Storage. There are basically three options in the related behavior of the import:

1. You don't specify the "os type" and you also don't specify that the imported image should be treated as "data disk" - In this scenario the GCP import job will try to inspect the image and if it finds out that the OS is supported, it will modify it on the fly by installing some GCP repositories and packages. If the OS is not supported (e.g. Fedora), then the import fails.
2. You specify the "os type" - GCP will inspect the image, but will disregard the OS type and treat it as if it was the specified OS type. Modification and installation of repos and packages is handled the same way as in case (1.)
3. You specify that the image is "data disk" - GCP treats this image as if it was not bootable and does not try to modify it. This does not have any other effect on the usage of the image, other than that GCP won't be able to inject some information into it (like create user, inject SSH keys, ...) and won't be able to use RHUI infrastructure inside GCP.

Based on the discussion we had, we agreed that we don't want GCP to modify our images for now and that it is OK to expect that the image had user (with set SSH keys) created as part of the image build. So as a result, the code now imports images build through cloudapi always as "data disks".

My motivation to not expose the "os" in the cloudapi was to keep it minimal for now, so that we don't need to worry about backward compatibility and API stability if we decide to remove it or not use it. But other than that, it could be exposed if API stability is not really important for now...

File: internal/cloudapi/server.go
Comment (thozza) [reply internal/cloudapi/server.go:238]: The problem from my point of view is that the image name must be unique within the GCP Project that we will use on our side for importing and sharing the image. Since the user won't see what image names we already have within the project, they could easily choose an image name that already exists. The unfortunate thing is that the workflow will fail quite late while importing the already uploaded image. So there is currently no real feedback when creating the compose request if the chosen image name already exists.

Nevertheless the internal GCP API could be extended with a function which would check if an image with the given name already exists within the GCP Project. But there could be race conditions (especially when more users use the API simultaneously) and also in order to expose this in cloudapi, the check would probably need to be done by a worker, which has GCP credentials.

So since I don't see any ideal solution here, I'll incline to whatever you think will be OK for the user...

Comment (teg) [reply internal/cloudapi/server.go:238]: We could leave this up to the caller (`image-builder` and make them responsible for avoiding clashes), but they'd basically be left with the same problem of having to pick a random name, they couldn't actually expose it to the user for the reasons you point out. I'd be ok with keeping it exactly as you have it here, then either `image-builder` would provide a name (appropriately randomized) or we'd create a random one internally.

File: internal/cloudapi/openapi.yml
Comment (teg) [reply internal/cloudapi/openapi.yml:268]: Makes sense! Let's discuss how to extend this after the first version is done.

File: tools/provision.sh
Comment (thozza) [reply tools/provision.sh:18]: Yeah, that probably makes sense. Both composer and worker configurations are currently stored in `test/data/composer` directory in git and installed to this directory.

Comment (teg) [reply tools/provision.sh:18]: Right, that seems like a mistake. Should have been split all the way.

File: test/cases/api.sh
Comment (thozza) [reply test/cases/api.sh:414]: Yes, this should be definitely added eventually. @Gundersanne also said that this should be good enough to start with. My expectation is that this may not be trivial, because GCP currently can not inject SSH keys into the instance which uses our imported image. This may require introducing the GCP image type.

File: test/README.md
Comment (thozza) [reply test/README.md:220]: Thanks for the note. I know about the containers work, but didn't use it much because it didn't work with `podman-compose` last time I tried it. And unfortunately I don't have docker installed, just podman. Nevertheless I definitely see how running the whole setup in separate containers could be beneficial for testing and debugging.

For now I mostly copied the wording from the composer-cli testing section which talked about running the test case locally. Personally I found running the cloudapi integration test directly on the GCP instance very useful, but it would probably run just OK in containers on the instance.

I may get back to the containers work and try to make it work with `podman-compose`.

**Comment by thozza** [comment]
> Seems we need some creds in Jenkins though. Do we have some we can use, or are we blocked on this?

I already added credentials to the Jenkins instance. I used the `cockpituous` project for now (CC @larskarlitski to expect some bills :smile: ). The error is actually a bug in the code, I'll fix it. The thing is that I currently made Jenkins to inject the credentials only for RHEL integration tests, that's why I used condition in the `provision.sh`.

File: test/cases/api.sh
Comment (teg) [reply test/cases/api.sh:414]: Makes sense.

File: internal/cloudapi/server.go
Comment (thozza) [reply internal/cloudapi/server.go:238]: @Gundersanne do you have any preference here? I basically did it the same way as AWS part does it. If the image name (snapshot name for AWS) was provided by the caller, just use it, otherwise use `composer-api-<UUID>`...

Comment (Gundersanne) [reply internal/cloudapi/server.go:238]: `fmt.Sprintf("composer-api-%s", uuid.New().String())` is fine for now I think. It's useful to have the default be some unique thing which is useful during testing.

**Review by ondrejbudai — COMMENTED** [review]
Looks great! Found two nits and one bigger thing for a discussion.

File: internal/cloudapi/server.go
Comment (thozza) [reply internal/cloudapi/server.go:238]: OK, thanks for all the comments. I'll just remove the "TODO" comment.

File: cmd/osbuild-upload-gcp/main.go
Comment (thozza) [reply cmd/osbuild-upload-gcp/main.go:58]: True, thanks!

File: internal/upload/gcp/gcp.go
Comment (teg) [reply internal/upload/gcp/gcp.go:220]: In general I think we should allow things to be cancelled by the caller. Not a fan of timeouts, except as a very last resort at the highest level (usually I'd say the user getting bored and clicking "cancel" is timeout good enough).
Might want to do the cancel logic in a follow-up though, seems like more invasive surgery would be needed.

Comment (ondrejbudai) [reply internal/upload/gcp/gcp.go:220]: Agreed!

Comment (thozza) [reply internal/upload/gcp/gcp.go:220]: Yeah, we probably should. I'm open to suggestions. The GCP build job has a timeout set to 7200 sec (2 hours) [on line 190](https://github.com/osbuild/osbuild-composer/pull/1245/files/cc4759b408a16325f4aeef7b9cc3efb4ef1714e2#diff-fa36b8cedd24680d43cde61fbdcd8681c6ccf4ebc452e394c34e692d575af6d3R190), so it will eventually fail. I could either expose this timeout or pass a context as you suggested. The magic value of 7200 sec is from GCP API example...

I didn't do it for the first iteration, because none of the internal cloud provider APIs allow to set timeout while they are still blocking.

Comment (teg) [reply internal/upload/gcp/gcp.go:220]: I'd keep it like this for now. Then go through in a follow-up and make sure all of  this (for all the providers) is cancelable.

File: tools/provision.sh
Comment (thozza) [reply tools/provision.sh]: This is fixed now.

File: cmd/osbuild-upload-gcp/main.go
Comment (thozza) [reply cmd/osbuild-upload-gcp/main.go]: This is fixed now.

File: internal/upload/gcp/gcp.go
Comment (thozza) [reply internal/upload/gcp/gcp.go]: This is fixed now.

File: test/cases/api.sh
Comment (ondrejbudai) [inline test/cases/api.sh:239]: I think you removed this step causing the CI to fail.

File: internal/cloudapi/openapi.yml
Comment (ondrejbudai) [inline internal/cloudapi/openapi.yml:259]: Just something that I've encountered right now: Can you document what happens if you don't pass the individual optional values?

File: test/cases/api.sh
Comment (thozza) [reply test/cases/api.sh:239]: Yeah, not sure how that happened. Thanks Ondrej!

File: schutzbot/Jenkinsfile
Comment (ondrejbudai) [inline schutzbot/Jenkinsfile:14]: I wonder if we should move these two variables in `test/cases/api.sh`. IMHO, it would be more convenient to have the tests as much self-contained as possible and don't put additional login in Jenkins.

Comment (thozza) [reply schutzbot/Jenkinsfile:14]: Personally I don't think that we should. The reason I see is that the value is specific to the testing environment used for running the test case. This bucket is not auto-created by the test, therefore the test case script would be not usable out of the box if you would not have access to the same project that we use for CI. Also Bucket name must be unique globally (among all GCP projects), so user can not create it in their own project.

File: internal/cloudapi/openapi.yml
Comment (thozza) [reply internal/cloudapi/openapi.yml:259]: I added description to each value with explanation of what will happen.

File: test/cases/api.sh
Comment (thozza) [reply test/cases/api.sh:239]: This is fixed now...

**Comment by msehnout** [comment]
Should we also extend cloud-cleaner with support for GCP? It is supposed to clean up any artifacts left behind in case of test failure so it the new test cases upload anything to GCP, we probably should.

**Comment by thozza** [comment]
> Should we also extend cloud-cleaner with support for GCP? It is supposed to clean up any artifacts left behind in case of test failure so it the new test cases upload anything to GCP, we probably should.

That's a fair point. There are currently code paths when the uploaded image can stay in the storage if something failed. I would rather refactor the internal API code a bit to ensure that the proper cleanup is done in case of an error when using the API. The service should not leave behind any leftover files in case of fail in production, where `cloud-cleaner` would not be run. There may be a failure with GCP when the upload succeeds, but image import fails because an image with chosen name already exists.

I'm happy to do it as part of this PR or create an issue and follow up on this...

**Comment by teg** [comment]
Good point @msehnout, I think landing this first and doing a follow-up is ok though.

**Comment by msehnout** [comment]
> > Should we also extend cloud-cleaner with support for GCP? It is supposed to clean up any artifacts left behind in case of test failure so it the new test cases upload anything to GCP, we probably should.
> 
> That's a fair point. There are currently code paths when the uploaded image can stay in the storage if something failed. I would rather refactor the internal API code a bit to ensure that the proper cleanup is done in case of an error when using the API. The service should not leave behind any leftover files in case of fail in production, where `cloud-cleaner` would not be run. There may be a failure with GCP when the upload succeeds, but image import fails because an image with chosen name already exists.
> 

The cloud cleaner is targeted at cases where Jenkins kills the job for some reason so any cleanup code path is useless. That's why it runs in the
```
post {
                        always {
```
section.

(Ideally it would also run as part of the cron job, but noone has written the code for querying images and removing anything older than whatever timeout we decide to use)

> I'm happy to do it as part of this PR or create an issue and follow up on this...

As Tom said, follow-up is fine :-)

File: internal/cloudapi/openapi.yml
Comment (Gundersanne) [inline internal/cloudapi/openapi.yml:305]: nitpick: why the newline here?

I'm not opposed to introducing newlines to make it easier to read but per `schemas` entry then

Comment (thozza) [reply internal/cloudapi/openapi.yml:305]: I didn't want the case "if not specified" to blend with the list. So just a personal preference. Happy to remove it if you think that it should not be there :wink:

Comment (thozza) [reply internal/cloudapi/openapi.yml]: I removed the extra newline...

File: tools/provision.sh
Comment (ondrejbudai) [inline tools/provision.sh:27]: We might even use a static filename.

Comment (thozza) [reply tools/provision.sh:27]: We can. I probably don't have a strong preference.

Comment (ondrejbudai) [reply tools/provision.sh:27]: Me neither, I'm fine with this. :)

Comment (thozza) [reply tools/provision.sh]: I changed it :smile:

**Review by teg — APPROVED** [review]
Amazing work! Thanks so much for this!

**Comment by thozza** [comment]
> I still need to verify that the changes don't affect `cockpit-composer`, but feel free to review in the meantime. Thanks!

Since no new image type has been added for GCP as well as no Weldr API support, there was no effect on `cockpit-composer` or `composer-cli` from what I tested. In case there are no outstanding issues that would be blocking this PR and CI passes, this change should be OK to merge. Thanks to all for a great feedback and discussion!

**Review by ondrejbudai — APPROVED** [review]
Thanks!

Amazing!

**Pull request merged**
